executor/cook/progress.py

```diff
@@ -71,37 +71,42 @@ def send_progress_update(self, progress_data, force_send=False):
         Nothing 
         """
         with self.lock:
-            if progress_data is not None and self.last_progress_data_sent != progress_data:
-                if force_send or self.has_enough_time_elapsed_since_last_update():
-                    logging.info('Sending progress message {}'.format(progress_data))
-                    message_dict = dict(progress_data)
-                    message_dict['task-id'] = self.task_id
-
-                    raw_progress_message = progress_data['progress-message']
-                    try:
-                        progress_str = raw_progress_message.decode('ascii').strip()
-                    except UnicodeDecodeError:
-                        logging.info('Unable to decode progress message in ascii, using empty string instead')
-                        progress_str = ''
-                    message_dict['progress-message'] = progress_str
-
-                    progress_message = json.dumps(message_dict)
-                    if len(progress_message) > self.max_message_length:
-                        num_extra_chars = len(progress_message) - self.max_message_length
-                        allowed_progress_message_length = max(len(progress_str) - num_extra_chars - 3, 0)
-                        new_progress_str = progress_str[:allowed_progress_message_length].strip() + '...'
-                        logging.info('Progress message trimmed to {}'.format(new_progress_str))
-                        message_dict['progress-message'] = new_progress_str
-                        progress_message = json.dumps(message_dict)
+            last_progress_data_sent = self.last_progress_data_sent
+            if progress_data is not None and last_progress_data_sent != progress_data:
+                last_progress_sequence = last_progress_data_sent['progress-sequence'] if last_progress_data_sent else 0
+                if last_progress_sequence < progress_data['progress-sequence']:
```

Comment:
What is this helping beyond the `self.has_enough_time_elapsed_since_last_update()` condition? Are the sequence numbers not always increasing? It seems like the equals check covers this already.

Reply:
There is force send at the end from each of the three (or two) `ProgressTracker`s and the equals check will fail if the sequence numbers do not match. We perform the secondary sequence check to ensure we do not send an outdated progress message.

Reply:
I think this is worth adding a comment.

Reply:
Done.